V2 content (#357)
enviodev/docs: docs/HyperIndex/Guides/event-handlers.mdx

in the project directory to generate the functions you will use in your handlers.

> Each event requires that a handler be registerred. These functions are asyncronous.

Reviewer: I don't think we require to register an event. It's just not processed when a handler is not registered.

Author: Hmm, yeah, you're right. I'll re-word slightly.

```suggestion
Each event that you want to process requires that a handler be registered. These functions are asynchronous.
```

> **Note:** _Two other functions are also sometimes used but not required. Both are explored in the Advanced section. One is used to optimise the loading of entities from the database in performance critical situations, the other is to be used for dynamic/factory contracts loading._

Reviewer: > **Note:** _Two other functions are also sometimes used but not required.

It's not very clear what functions you're talking about here.

Author: ```suggestion
**Note:** _Two other functions, `handlerWithLoader` and `contractRegister`, are also sometimes used but not required for most indexers. Both functions are explored in the Advanced section. `handlerWithLoader` is used to optimise the loading of entities from the database in performance critical situations, `contractRegister` is to be used for dynamic/factory contracts loading._
<!-- TODO: add links to the advanced section about handlerWithLoader and contractRegister -->
```

## Handler function

Handler functions are called via

<Tabs>
  <TabItem value="typescript" label="Typescript">

```typescript
import { <ContractName> } from "generated";

> <ContractName>.< EventName >.handler(async ({event, context}) => {

Reviewer: ```suggestion
<ContractName>.<EventName>.handler(async ({event, context}) => {
```

  // Your logic
});
```

  </TabItem>
  <TabItem value="javascript" label="Javascript">

```javascript

> const <ContractName> = require("generated").<ContractName>;

Reviewer: It's more flexible to extend:
```suggestion
const { <ContractName> } = require("generated");
```

> <ContractName>.< EventName >.handler(async ({event, context}) => {

Reviewer: ```suggestion
<ContractName>.<EventName>.handler(async ({event, context}) => {
```

  // Your logic
});
```

  </TabItem>
  <TabItem value="rescript" label="Rescript">

```rescript

> <ContractName>Contract.<EventName>.handler(async ({event, context}) => {

Reviewer: ```suggestion
Handlers.<ContractName>.<EventName>.handler(async ({event, context}) => {
```

Author: Good catch!

### Context

In your handler functions you can access entities in the database via the asyncronous 'get' function as shown below:

```javascript
await context.<EntityName>.get(<id>)
```

> The context also contains some logging functions that include the context automatically. You can also use `console.log`.

Reviewer: For me, it's not clear why I should use the logging functions on the context instead of `console.log.` And an example would be nice.

Author: ```suggestion
The context also contains some logging functions that include the context to improve the debugging experience, such as the block and event that is currently being processed automatically. You can also use `console.log` if you don't benefit from that extra context.
```

context.<EntityName>.deleteUnsafe(<entityObject>.id)
```

> The set function is used to either create an entity or to update an existing entity with the values defined in the `entityObject`.

Reviewer: ```suggestion
The `set` method is used to either create an entity or to update an existing entity with the values defined in the `entityObject`.
```

> The deleteUnsafe function is an experimental feature that allows you to remove an entity with a particular ID from the database and indexer memory. This is un `unsafe` function,

Reviewer: ```suggestion
The `deleteUnsafe` method is an experimental feature that allows you to remove an entity with a particular ID from the database and indexer memory. This is an **unsafe** function,
```

  <TabItem value="typescript" label="Typescript">

```typescript
import { Greeter, User } from "generated";

> Greeter.NewGreeting_handler(async ({ event, context }) => {

Reviewer: ```suggestion
Greeter.NewGreeting.handler(async ({ event, context }) => {
```

```rescript
open Types

Handlers.Greeter.NewGreeting.handler(({ event, context }) => {
  let currentUserOpt = context.user.get(event.params.user->Ethers.ethAddressToString)

  switch currentUserOpt {
  | Some(existingUser) => {

>       let userObject: User = {

Reviewer: ```suggestion
      let userObject: user = {
```

Author: Ah, good catch!

    }

  | None =>

>     let userObject: User = {

Reviewer: ```suggestion
    let userObject: user = {
```

```typescript
import { getConfigByChainId } from "../generated/src/ConfigYAML.gen";

> Greeter_NewGreeting_handler(async ({ event, context }) => {

Reviewer: ```suggestion
Greeter.NewGreeting.handler(async ({ event, context }) => {
```
